### DISCUSSION TITLE
【提案】更有语义化的对数组元素删除或插入的方法

### DISCUSSION THREAD

**Discussion by otomad** [JSCIG/es-discuss#65] [Ideas]
其他地方似乎也有不少讨论如：
https://es.discourse.group/t/array-prototype-remove/760

JavaScript 虽然有一个 `Array.prototype.splice` 函数可以实现这些功能且更强大，但是它对新手或其他语言转来的人不直观。因此拟提出以下方法：

## `Array.prototype.removeAt`
从数组的指定索引值处删除元素，将会改变原始数组。返回是否成功删除的布尔值。
#### 定义
```typescript
Array.prototype.removeAt = function<T>(index: number): boolean {
    const removed = this.splice(index, 1);
    return !!removed.length;
}
```

## `Array.prototype.remove`
从数组中删除指定的元素，将会改变原始数组。返回成功删除元素的个数。
#### 注意
1. 数组中若有多个重复的该元素，只删除第一个该元素。
2. 可以在函数的参数列表中同时提供多个要删除的元素。
#### 定义
```typescript
Array.prototype.remove = function<T>(...items: T[]): number {
    let removedCount = 0;
    for (const item of items) {
        const index = this.indexOf(item);
        if (index === -1) continue;
        this.splice(index, 1);
        removedCount++;
    }
    return removedCount;
}
```

## `Array.prototype.insert`
将新元素插入到数组的指定索引处，将会改变原始数组。
#### 注意：
1. 可以在函数的参数列表中同时提供多个要插入的元素。
2. 插入的索引值可以指定为负值，如 -1 表示插入到最后。
#### 定义
```typescript
Array.prototype.insert = function<T>(index: number, ...items: T[]): void {
    if (index < 0) index = this.length + 1 + index;
    this.splice(index, 0, ...items);
}
```

----

同时，由于现在有一个已经结案了的通过复制更改数组的提案，因此也可以考虑同时添加以下方法，它们将不改变原始数组，而是创建原始数组的浅拷贝，并返回操作后的数组副本。
```typescript
Array.prototype.toRemovedAt = function<T>(index: number): T[];
Array.prototype.toRemoved = function<T>(...items: T[]): T[];
Array.prototype.toInserted = function<T>(index: number, ...items: T[]): T[];
```

**Comment by ljharb** [comment]
I don't think it's likely that a new method that mutates arrays will ever be added, at this point, and since `spliced` and `toSpliced` already cover this functionality, I'm not sure if it's worth adding these methods.

**Reply by otomad** [reply]
Except `remove` item, you should get the index of the item first, then use `splice` to delete the index of the item.

**Reply by ljharb** [reply]
True. Altho for that one you could be doing `const toRemove = new Set(items); arr.filter(x => !toRemove.has(x))`
